middleware/passport/githubStrategy.ts

```diff
@@ -0,0 +1,121 @@
+  const config = app.settings.runtimeConfig;
+  const providers = app.settings.providers as IProviders;
+  if (config && config.impersonation && config.impersonation.githubId) {
+    const operations = providers.operations as Operations;
+    const impersonationId = config.impersonation.githubId;
+
+    const account = operations.getAccount(impersonationId);
+    return account.getDetails().then(details => {
+      console.warn(`GITHUB IMPERSONATION: id=${impersonationId} login=${details.login} name=${details.name}`);
+      return done(null, {
+        github: {
+          accessToken: 'fakeaccesstoken',
+          displayName: details.name,
+          avatarUrl: details.avatar_url,
+          id: details.id.toString(),
+          username: details.login,
+        },
+      });
+    }).catch(err => {
+      return done(err);
+    });
+  }
+  let subset = {
+    github: {
+      accessToken: accessToken,
+      displayName: profile.displayName,
+      avatarUrl: profile._json && profile._json.avatar_url ? profile._json.avatar_url : undefined,
+      id: profile.id,
+      username: profile.username,
+      scope: undefined,
+    },
+  };
+  if (modernAppInUse) {
+    subset.github.scope = 'githubapp';
+  }
+  return done(null, subset);
+}
+
+function githubResponseToIncreasedScopeSubset(modernAppInUse: boolean, accessToken: string, refreshToken: string, profile, done) {
+  if (modernAppInUse) {
+    return done(new Error('githubResponseToIncreasedScopeSubset is not compatible with modern apps'));
+  }
+  const subset = {
+    githubIncreasedScope: {
+      accessToken: accessToken,
+      id: profile.id,
+      username: profile.username,
+    },
+  };
+  return done(null, subset);
+}
+
+export function getGitHubAppConfigurationOptions(config) {
+  let legacyOAuthApp = config.github.oauth2 && config.github.oauth2.clientId && config.github.oauth2.clientSecret ? config.github.oauth2 : null;
+  const customerFacingApp = config.github.app && config.github.app.ui && config.github.app.ui.clientId && config.github.app.ui.clientSecret ? config.github.app.ui : null;
+  const useCustomerFacingGitHubAppIfPresent = config.github.oauth2.useCustomerFacingGitHubAppIfPresent === true;
+  if (useCustomerFacingGitHubAppIfPresent && customerFacingApp) {
+    if (legacyOAuthApp && legacyOAuthApp['callbackUrl']) {
+      customerFacingApp['callbackUrl'] = legacyOAuthApp['callbackUrl'];
+    }
+    legacyOAuthApp = null;
+  }
+  const modernAppInUse: boolean = customerFacingApp && !legacyOAuthApp;
+  const githubAppConfiguration = modernAppInUse ? customerFacingApp : legacyOAuthApp;
+  return { legacyOAuthApp, customerFacingApp, modernAppInUse, githubAppConfiguration };
+}
+
+export default function createGitHubStrategy(app, config) {
+  let strategies = {};
+  const { modernAppInUse, githubAppConfiguration } = getGitHubAppConfigurationOptions(config);
+  // NOTE: due to bugs in the GitHub API v3 around user-to-server requests in
```

Comment:
This `NOTE` can finally be removed. I'm happy to do it separately but thought I'd note that here.